pkg/shell/nav.jsx

```diff
@@ -0,0 +1,201 @@
+                i += step;
+                if (i < 0 || i >= all.length)
+                    document.querySelector("#" + sel + " .filter-menus").focus();
+                else
+                    all[i].focus();
+            }
+        }
+
+        function navigate_apps(ev) {
+            if (ev.keyCode === 13) // Enter
+                clickActiveItem();
+            else if (ev.keyCode === 40) // Arrow Down
+                focusNextItem(0, 1);
+            else if (ev.keyCode === 38) // Arrow Up
+                focusNextItem(-1, -1);
+            else if (ev.keyCode === 27) { // Escape - clean selection
+                self.setState({ search: "" });
+                document.querySelector("#" + sel + " .filter-menus").focus();
+            }
+        }
+
+        document.getElementById(sel).addEventListener("keyup", navigate_apps);
+        document.getElementById(sel).addEventListener("change", navigate_apps);
+    }
+
+    static getDerivedStateFromProps(nextProps, prevState) {
+        if (nextProps.current !== prevState.current)
+            return {
+                current: nextProps.current,
+                search: "",
+            };
+        return null;
+    }
+
+    onSearchChanged(e) {
+        this.setState({ search: e.target.value });
+    }
+
+    clearSearch() {
+        this.setState({ search: "" });
+    }
+
+    render() {
+        const groups = [];
+        const term = this.state.search.toLowerCase();
+        this.props.groups.forEach(g => {
+            const new_items = g.items.map(i => this.props.filtering(i, term)).filter(Boolean);
+            new_items.sort(this.props.sorting);
+            if (new_items.length > 0)
+                groups.push({ name: g.name, items: new_items });
+        });
+
+        return (
+            <>
+                <div className="has-feedback search">
+                    <input className="filter-menus form-control" type="search" placeholder={_("Search")} aria-label={_("Search")} onChange={this.onSearchChanged} value={this.state.search} />
+                    <span className="fa fa-search form-control-feedback" />
+                </div>
+                <Nav onSelect={this.onSelect} theme="dark">
+                    { groups.map(g =>
+                        <NavGroup key={g.name} title={g.name}>
+                            {g.items.map(i => this.props.item_render(i, this.state.search.toLowerCase()))}
+                        </NavGroup>
+                    )}
+                    { groups.length < 1 && <span className="non-menu-item">{_("No results found")}</span> }
+                    { this.state.search !== "" && <span className="non-menu-item"><button onClick={this.clearSearch} className="link-button hint">{_("Clear Search")}</button></span> }
+                </Nav>
+            </>
+        );
+    }
+}
+
+CockpitNav.propTypes = {
+    groups: PropTypes.array,
+    selector: PropTypes.string,
+    item_render: PropTypes.func,
```

Comment:
At least this property and groups are  `.isRequired`, and probably a few others.